File: csharp/AppEncryption/AppEncryption/Persistence/DynamoDbMetastoreImpl.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.CompilerServices;
using System.Threading.Tasks;
using Amazon;
using Amazon.DynamoDBv2;
using Amazon.DynamoDBv2.DocumentModel;
using Amazon.DynamoDBv2.Model;
using Amazon.Runtime;
using App.Metrics.Timer;
using GoDaddy.Asherah.AppEncryption.Util;
using GoDaddy.Asherah.Crypto.Exceptions;
using GoDaddy.Asherah.Logging;
using LanguageExt;
using Microsoft.Extensions.Logging;
using Newtonsoft.Json.Linq;

[assembly: InternalsVisibleTo("DynamicProxyGenAssembly2")]
[assembly: InternalsVisibleTo("AppEncryption.Tests")]

namespace GoDaddy.Asherah.AppEncryption.Persistence
{
    /// <summary>
    /// Provides an AWS DynamoDB based implementation of <see cref="IMetastore{T}"/> to store and retrieve system keys
    /// and intermediate keys as <see cref="JObject"/> values. These are used by Asherah to provide a hierarchical key
    /// structure. It uses the default table name "EncryptionKey" but it can be configured using
    /// <see cref="IBuildStep.WithTableName"/> option. Stores the created time in unix time seconds.
    /// </summary>
    public class DynamoDbMetastoreImpl : IMetastore<JObject>
    {
        internal const string PartitionKey = "Id";
        internal const string SortKey = "Created";
        internal const string AttributeKeyRecord = "KeyRecord";

        private static readonly TimerOptions LoadTimerOptions = new TimerOptions { Name = MetricsUtil.AelMetricsPrefix + ".metastore.dynamodb.load" };
        private static readonly TimerOptions LoadLatestTimerOptions = new TimerOptions { Name = MetricsUtil.AelMetricsPrefix + ".metastore.dynamodb.loadlatest" };
        private static readonly TimerOptions StoreTimerOptions = new TimerOptions { Name = MetricsUtil.AelMetricsPrefix + ".metastore.dynamodb.store" };

        private static readonly ILogger Logger = LogManager.CreateLogger<DynamoDbMetastoreImpl>();

        private static readonly string DefaultKeySuffix = string.Empty;
        private static bool hasKeySuffix;
        private readonly string preferredRegion;
        private readonly Table table;

        internal DynamoDbMetastoreImpl(Builder builder)
        {
            DbClient = builder.DbClient;
            TableName = builder.TableName;
            preferredRegion = builder.PreferredRegion;
            hasKeySuffix = builder.HasKeySuffix;

            // Note this results in a network call. For now, cleaner than refactoring w/ thread-safe lazy loading
            table = builder.LoadTable(DbClient, TableName);
        }

        public interface IBuildStep
        {
            /// <summary>
            /// Specifies whether key suffix should be enabled for DynamoDB.
            /// </summary>
            ///
            /// <returns>The current <see cref="IBuildStep"/> instance.</returns>
            IBuildStep WithKeySuffix();

            /// <summary>
            /// Used to define a custom table name in DynamoDB. The default name is
            /// <see cref="Builder.DefaultTableName"/>.
            /// </summary>
            ///
            /// <param name="tableName">The custom table name to use.</param>
            /// <returns>The current <see cref="IBuildStep"/> instance.</returns>
            IBuildStep WithTableName(string tableName);

            /// <summary>
            /// Used to define custom AWS credentials for the DynamoDB client.
            /// </summary>
            ///
            /// <param name="credentials">The custom AWS credentials to use.</param>
            /// <returns>The current <see cref="IBuildStep"/> instance.</returns>
            IBuildStep WithCredentials(AWSCredentials credentials);

            /// <summary>
            /// Adds Endpoint config to the AWS DynamoDb client.
            /// </summary>
            ///
            /// <param name="endPoint">the service endpoint either with or without the protocol.</param>
            /// <param name="signingRegion">the region to use for SigV4 signing of requests (e.g. us-west-1).</param>
            /// <returns>The current <see cref="IBuildStep"/> instance.</returns>
            IBuildStep WithEndPointConfiguration(string endPoint, string signingRegion);

            /// <summary>
            /// Specifies the region for the AWS DynamoDb client. If this is not specified, then the region from
            /// <see cref="DynamoDbMetastoreImpl.NewBuilder"/> is used.
            /// </summary>
            ///
            /// <param name="region">The region for the DynamoDb client.</param>
            /// <returns>The current <see cref="IBuildStep"/> instance.</returns>
            IBuildStep WithRegion(string region);

            /// <summary>
            /// Builds the finalized <see cref="DynamoDbMetastoreImpl"/> with the parameters specified in the builder.
            /// </summary>
            ///
            /// <returns>The fully instantiated <see cref="DynamoDbMetastoreImpl"/> object.</returns>
            DynamoDbMetastoreImpl Build();
        }

        internal string TableName { get; }

        internal IAmazonDynamoDB DbClient { get; }

        /// <summary>
        /// Builder method to initialize a <see cref="DynamoDbMetastoreImpl"/> instance.
        /// </summary>
        ///
        /// <param name="region">AWS region to use with the DynamoDB implementation.</param>
        /// <returns>The current <see cref="Builder"/> instance.</returns>
        public static Builder NewBuilder(string region)
        {
            return new Builder(region);
        }

        /// <inheritdoc />
        public Option<JObject> Load(string keyId, DateTimeOffset created)
        {
            using (MetricsUtil.MetricsInstance.Measure.Timer.Time(LoadTimerOptions))
            {
                try
                {
                    GetItemOperationConfig config = new GetItemOperationConfig
                    {
                        AttributesToGet = new List<string> { AttributeKeyRecord },
                        ConsistentRead = true, // Always use strong consistency
                    };
                    Task<Document> getItemTask = table.GetItemAsync(keyId, created.ToUnixTimeSeconds(), config);
                    Document result = getItemTask.Result;
                    if (result != null)
                    {
                        // TODO Optimize Document to JObject conversion. Helper method could enumerate over Document KeyPairs
                        // and convert DynamoDBEntry values based on type inspection
                        return Option<JObject>.Some(JObject.Parse(result[AttributeKeyRecord].AsDocument().ToJson()));
                    }
                }
                catch (AggregateException ae)
                {
                    Logger.LogError(ae, "Metastore error");
                }

                return Option<JObject>.None;
            }
        }

        /// <summary>
        /// Lookup the latest value associated with the keyId.
        /// </summary>
        ///
        /// <param name="keyId">The keyId to lookup.</param>
        /// <returns>The latest <seealso cref="JObject"/> value associated with the keyId, if any.</returns>
        public Option<JObject> LoadLatest(string keyId)
        {
            using (MetricsUtil.MetricsInstance.Measure.Timer.Time(LoadLatestTimerOptions))
            {
                // Have to use query api to use limit and reverse sort order
                try
                {
                    QueryFilter filter = new QueryFilter(PartitionKey, QueryOperator.Equal, keyId);
                    QueryOperationConfig config = new QueryOperationConfig
                    {
                        Limit = 1,
                        ConsistentRead = true, // always use strong consistency
                        BackwardSearch = true, // sorts descending
                        Filter = filter,
                        AttributesToGet = new List<string> { AttributeKeyRecord },
                        Select = SelectValues.SpecificAttributes,
                    };
                    ISearch search = table.Query(config);
                    Task<List<Document>> searchTask = search.GetNextSetAsync();
                    List<Document> result = searchTask.Result;
                    if (result.Count > 0)
                    {
                        Document keyRecordDocument = result.First();

                        // TODO Optimize Document to JObject conversion. Helper method could enumerate over Document KeyPairs
                        // and convert DynamoDBEntry values based on type inspection
                        return Option<JObject>.Some(JObject.Parse(keyRecordDocument[AttributeKeyRecord].AsDocument().ToJson()));
                    }
                }
                catch (AggregateException se)
                {
                    Logger.LogError(se, "Metastore error");
                }

                return Option<JObject>.None;
            }
        }

        /// <summary>
        /// Stores the <see cref="JObject"/> value using the specified keyId and created time.
        /// </summary>
        ///
        /// <param name="keyId">The keyId to store.</param>
        /// <param name="created">The created time to store.</param>
        /// <param name="value">The <see cref="JObject"/> value to store.</param>
        /// <returns><value>True</value> if the store succeeded, <value>False</value> if the store failed for a known
        /// condition like duplicate key.</returns>
        /// <exception cref="AppEncryptionException">Raise an exception in case of any other metastore errors.
        /// </exception>
        public bool Store(string keyId, DateTimeOffset created, JObject value)
        {
            using (MetricsUtil.MetricsInstance.Measure.Timer.Time(StoreTimerOptions))
            {
                try
                {
                    Document document = new Document
                    {
                        [PartitionKey] = keyId,
                        [SortKey] = created.ToUnixTimeSeconds(),

                        // TODO Optimize JObject to Document conversion. Just need lambda that calls Document.
                        // Add and recurses for Dictionary and List types
                        [AttributeKeyRecord] = Document.FromJson(value.ToString()),
                    };

                    // Note conditional expression using attribute_not_exists has special semantics. Can be used on
                    // partition OR sort key alone to guarantee primary key uniqueness. It automatically checks for
                    // existence of this item's composite primary key and if it contains the specified attribute name,
                    // either of which is inherently required.
                    Expression expr = new Expression
                    {
                        ExpressionStatement = "attribute_not_exists(" + PartitionKey + ")",
                    };
                    PutItemOperationConfig config = new PutItemOperationConfig
                    {
                        ConditionalExpression = expr,
                    };

                    table.PutItemAsync(document, config).Wait();
                    return true;
                }
                catch (AggregateException ae)
                {
                    foreach (Exception exception in ae.InnerExceptions)
                    {
                        if (exception is ConditionalCheckFailedException)
                        {
                            Logger.LogInformation("Attempted to create duplicate key: {KeyId} {Created}", keyId, created);
                            return false;
                        }
                    }

                    Logger.LogError(ae, "Metastore error during store");
                    throw new AppEncryptionException("Metastore error:", ae);
                }
            }
        }

        public string GetKeySuffix()
        {
            if (hasKeySuffix)
            {
                return preferredRegion;
            }

            return DefaultKeySuffix;
        }

        /// <summary>
        /// Builder class to create an instance of the <see cref="DynamoDbMetastoreImpl"/> class.
        /// </summary>
        public class Builder : IBuildStep, IDisposable
        {
            private readonly string preferredRegion;
            private AmazonDynamoDBClient dbClient;
            private bool hasKeySuffix;
            private string tableName = DefaultTableName;
            private AWSCredentials credentials;

            // Internal properties for access
            internal string PreferredRegion => preferredRegion;
            internal AmazonDynamoDBClient DbClient => dbClient;
            internal bool HasKeySuffix => hasKeySuffix;
            internal string TableName => tableName;
            internal AWSCredentials Credentials => credentials;

            private const string DefaultTableName = "EncryptionKey";
            private readonly AmazonDynamoDBConfig dbConfig = new AmazonDynamoDBConfig();
            private bool hasEndPoint;
            private bool hasRegion;

            /// <summary>
            /// Initializes a new instance of the <see cref="Builder"/> class.
            /// </summary>
            /// <param name="region">The region to use with AWS DynamoDB.</param>
            public Builder(string region)
            {
                preferredRegion = region;
            }

            /// <inheritdoc/>
            public IBuildStep WithKeySuffix()
            {
                hasKeySuffix = true;
                return this;
            }

            /// <inheritdoc/>
            public IBuildStep WithTableName(string tableName)
            {
                this.tableName = tableName;
                return this;
            }

            /// <inheritdoc/>
            public IBuildStep WithEndPointConfiguration(string endPoint, string signingRegion)
            {
                if (!hasRegion)
                {
                    hasEndPoint = true;
                    dbConfig.ServiceURL = endPoint;
                    dbConfig.AuthenticationRegion = signingRegion;
                }

                return this;
            }

            /// <inheritdoc/>
            public IBuildStep WithRegion(string region)
            {
                if (!hasEndPoint)
                {
                    hasRegion = true;
                    dbConfig.RegionEndpoint = RegionEndpoint.GetBySystemName(region);
                }

                return this;
            }

            /// <inheritdoc/>
            public IBuildStep WithCredentials(AWSCredentials credentials)
            {
                this.credentials = credentials;
                return this;
            }

            /// <summary>
            /// Builds the finalized <see cref="DynamoDbMetastoreImpl"/> object with the parameters specified in the
            /// <see cref="Builder"/>.
            /// </summary>
            ///
            /// <returns>The fully instantiated <see cref="DynamoDbMetastoreImpl"/> object.</returns>
            public DynamoDbMetastoreImpl Build()
            {
                if (!hasEndPoint && !hasRegion)
                {
                    dbConfig.RegionEndpoint = RegionEndpoint.GetBySystemName(preferredRegion);
                }

                dbClient = new AmazonDynamoDBClient(credentials, dbConfig);

                return new DynamoDbMetastoreImpl(this);
            }

            internal virtual Table LoadTable(IAmazonDynamoDB client, string tableName)
            {
                return new TableBuilder(client, tableName)
                    .AddHashKey(PartitionKey, DynamoDBEntryType.String)
                    .AddRangeKey(SortKey, DynamoDBEntryType.Numeric)
                    .Build();
            }

            /// <summary>
            /// Disposes of the managed resources.
            /// </summary>
            public void Dispose()
            {
                Dispose(true);
                GC.SuppressFinalize(this);
            }

            /// <summary>
            /// Disposes of the managed resources.
            /// </summary>
            /// <param name="disposing">True if called from Dispose, false if called from finalizer.</param>
            protected virtual void Dispose(bool disposing)
            {
                if (disposing)
                {
                    dbClient?.Dispose();
                }
            }
        }
    }
}

File: java/app-encryption/src/main/java/com/godaddy/asherah/appencryption/persistence/DynamoDbMetastoreImpl.java
package com.godaddy.asherah.appencryption.persistence;

import java.net.URI;
import java.time.Instant;
import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;
import java.util.Optional;

import software.amazon.awssdk.core.exception.SdkException;
import software.amazon.awssdk.regions.Region;
import software.amazon.awssdk.services.dynamodb.DynamoDbClient;
import com.godaddy.asherah.appencryption.exceptions.AppEncryptionException;
import com.godaddy.asherah.appencryption.utils.MetricsUtil;

import io.micrometer.core.instrument.Metrics;
import io.micrometer.core.instrument.Timer;

import org.json.JSONObject;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import software.amazon.awssdk.services.dynamodb.DynamoDbClientBuilder;
import software.amazon.awssdk.services.dynamodb.model.AttributeValue;
import software.amazon.awssdk.services.dynamodb.model.ComparisonOperator;
import software.amazon.awssdk.services.dynamodb.model.Condition;
import software.amazon.awssdk.services.dynamodb.model.ConditionalCheckFailedException;
import software.amazon.awssdk.services.dynamodb.model.GetItemResponse;
import software.amazon.awssdk.services.dynamodb.model.QueryResponse;

/**
 * Provides an AWS DynamoDB based implementation of {@link Metastore} to store and retrieve
 * {@link com.godaddy.asherah.appencryption.utils.Json} values for system and intermediate keys. It uses the default
 * table name "EncryptionKey" but that can be configured using {@link Builder#withTableName(String)} " option.
 * Creation time is stored in unix time seconds.
 */
public class DynamoDbMetastoreImpl implements Metastore<JSONObject> {
  static final String DEFAULT_KEY_SUFFIX = "";

  private static final Logger logger = LoggerFactory.getLogger(DynamoDbMetastoreImpl.class);

  static final String PARTITION_KEY = "Id";
  static final String SORT_KEY = "Created";
  static final String ATTRIBUTE_KEY_RECORD = "KeyRecord";

  private final Timer loadTimer = Metrics.timer(MetricsUtil.AEL_METRICS_PREFIX + ".metastore.dynamodb.load");
  private final Timer loadLatestTimer =
      Metrics.timer(MetricsUtil.AEL_METRICS_PREFIX + ".metastore.dynamodb.loadlatest");
  private final Timer storeTimer = Metrics.timer(MetricsUtil.AEL_METRICS_PREFIX + ".metastore.dynamodb.store");

  private final DynamoDbClient dynamoDbClient;
  private final String tableName;
  private final String preferredRegion;
  private final boolean hasKeySuffix;


  /**
   * Initialize a {@code DynamoDbMetastoreImpl} builder with the given region.
   *
   * @param region The preferred region for the DynamoDb. This can be overridden using the
   * {@link Builder#withRegion(String)} builder step.
   * @return The current {@code Builder} instance.
   */
  public static Builder newBuilder(final String region) {
    return new Builder(region);
  }

  DynamoDbMetastoreImpl(final Builder builder) {
    this.dynamoDbClient = builder.dynamoDbClient;
    this.tableName = builder.tableName;
    this.preferredRegion = builder.preferredRegion;
    this.hasKeySuffix = builder.hasKeySuffix;
  }

  @Override
  public Optional<JSONObject> load(final String keyId, final Instant created) {
    return loadTimer.record(() -> {
      try {
        GetItemResponse getItemResponse = dynamoDbClient.getItem(request ->
            request
              .tableName(tableName)
              .key(Map.of(
                PARTITION_KEY, AttributeValue.fromS(keyId),
                SORT_KEY, AttributeValue.fromN(Long.toString(created.getEpochSecond()))
              ))
              .attributesToGet(ATTRIBUTE_KEY_RECORD)
              .consistentRead(true)  // always use strong consistency
        );
        Map<String, AttributeValue> item = getItemResponse.item();
        if (item != null && item.containsKey(ATTRIBUTE_KEY_RECORD)) {
          return Optional.of(toJSONObject(item.get(ATTRIBUTE_KEY_RECORD).m()));
        }
      }
      catch (SdkException se) {
        logger.error("Metastore error", se);
      }

      return Optional.empty();
    });
  }

  /**
   * Lookup the latest value associated with the keyId.
   *
   * @param keyId The keyId part of the lookup key.
   * @return The latest value associated with the keyId, if any.
   */
  @Override
  public Optional<JSONObject> loadLatest(final String keyId) {
    return loadLatestTimer.record(() -> {
      try {
        // Have to use query api to use limit and reverse sort order
        QueryResponse queryResponse = dynamoDbClient.query(request ->
            request
              .tableName(tableName)
              .keyConditions(Map.of(
                PARTITION_KEY, Condition.builder()
                  .attributeValueList(AttributeValue.fromS(keyId))
                  .comparisonOperator(ComparisonOperator.EQ)
                  .build()
                ))
              .attributesToGet(ATTRIBUTE_KEY_RECORD)
              .scanIndexForward(false) // sorts descending
              .limit(1) // limit 1
              .consistentRead(true) // limit 1
        );
        Iterator<Map<String, AttributeValue>> iterator = queryResponse.items().iterator();
        if (iterator.hasNext()) {
          Map<String, AttributeValue> item = iterator.next();
          return Optional.of(toJSONObject(item.get(ATTRIBUTE_KEY_RECORD).m()));
        }
      }
      catch (SdkException se) {
        logger.error("Metastore error", se);
      }

      return Optional.empty();
    });
  }

  /**
   * Stores the value using the specified keyId and created time.
   *
   * @param keyId The keyId part of the lookup key.
   * @param created The created time part of the lookup key.
   * @param value The value to store.
   * @return {@code true} if the store succeeded, false if the store failed for a known condition
   *         e.g., trying to save a duplicate value should return false, not throw an exception.
   */
  @Override
  public boolean store(final String keyId, final Instant created, final JSONObject value) {
    return storeTimer.record(() -> {
      try {
        // Note conditional expression using attribute_not_exists has special semantics. Can be used on partition OR
        // sort key alone to guarantee primary key uniqueness. It automatically checks for existence of this item's
        // composite primary key and if it contains the specified attribute name, either of which is inherently
        // required.
        dynamoDbClient.putItem(request ->
            request
              .tableName(tableName)
              .item(Map.of(
                PARTITION_KEY, AttributeValue.fromS(keyId),
                SORT_KEY, AttributeValue.fromN(Long.toString(created.getEpochSecond())),
                ATTRIBUTE_KEY_RECORD, AttributeValue.fromM(toDynamoDbItem(value))
              ))
              .conditionExpression("attribute_not_exists(" + PARTITION_KEY + ")")
        );
        return true;
      }
      catch (ConditionalCheckFailedException e) {
        // Duplicate key exists
        logger.info("Attempted to create duplicate key: {} {}", keyId, created);
        return false;
      }
      catch (SdkException se) {
        logger.error("Metastore error during store", se);
        throw new AppEncryptionException("Metastore error", se);
      }
    });
  }

  static Map<String, AttributeValue> toDynamoDbItem(final JSONObject value) {
    Map<String, AttributeValue> result = new HashMap<>();
    for (final var key : value.keySet()) {
      final var object = value.get(key);
      if (object instanceof String stringObject) {
        result.put(key, AttributeValue.fromS(stringObject));
      }
      else if (object instanceof Long longObject) {
        result.put(key, AttributeValue.fromN(Long.toString(longObject)));
      }
      else if (object instanceof Boolean booleanObject) {
        result.put(key, AttributeValue.fromBool(booleanObject));
      }
      else if (object instanceof JSONObject jsonObject) {
        result.put(key, AttributeValue.fromM(toDynamoDbItem(jsonObject)));
      }
      else {
        throw new IllegalArgumentException("Unsupported type: " + object.getClass().getName());
      }
    }
    return result;
  }

  static JSONObject toJSONObject(final Map<String, AttributeValue> item) {
    JSONObject result = new JSONObject();
    for (final var entry : item.entrySet()) {
      final var value = entry.getValue();
      if (value.s() != null) {
        result.put(entry.getKey(), value.s());
      }
      else if (value.n() != null) {
        result.put(entry.getKey(), Long.parseLong(value.n()));
      }
      else if (value.bool() != null) {
        result.put(entry.getKey(), value.bool());
      }
      else if (value.m() != null) {
        result.put(entry.getKey(), toJSONObject(value.m()));
      }
      else {
        throw new IllegalArgumentException("Unsupported type: " + value.type());
      }
    }
    return result;
  }

  @Override
  public String getKeySuffix() {
    if (hasKeySuffix) {
      return preferredRegion;
    }
    return DEFAULT_KEY_SUFFIX;
  }

  String getTableName() {
    return tableName;
  }

  DynamoDbClient getClient() {
    return dynamoDbClient;
  }

  public static final class Builder implements BuildStep, EndPointStep, RegionStep {
    static final String DEFAULT_TABLE_NAME = "EncryptionKey";

    private DynamoDbClient dynamoDbClient;

    private final String preferredRegion;
    private final DynamoDbClientBuilder clientBuilder = DynamoDbClient.builder();

    private String tableName = DEFAULT_TABLE_NAME;
    private boolean hasEndPoint = false;
    private boolean hasKeySuffix = false;
    private boolean hasRegion = false;

    private DynamoDbClient clientOverride;

    private Builder(final String region) {
      this.preferredRegion = region;
    }

    @Override
    public BuildStep withTableName(final String table) {
      this.tableName = table;
      return this;
    }

    @Override
    public BuildStep withEndPointConfiguration(final String endPoint, final String signingRegion) {
      if (!hasRegion) {
        hasEndPoint = true;
        clientBuilder
            .endpointOverride(URI.create(endPoint))
            .region(Region.of(signingRegion));
      }
      return this;
    }

    @Override
    public BuildStep withKeySuffix() {
      this.hasKeySuffix = true;
      return this;
    }

    @Override
    public BuildStep withRegion(final String region) {
      if (!hasEndPoint) {
        hasRegion = true;
        clientBuilder.region(Region.of(region));
      }
      return this;
    }

    @Override
    public BuildStep withClientOverride(final DynamoDbClient client) {
      this.clientOverride = client;
      return this;
    }

    @Override
    public DynamoDbMetastoreImpl build() {
      if (clientOverride != null) {
        dynamoDbClient = clientOverride;
        return new DynamoDbMetastoreImpl(this);
      }
      if (!hasEndPoint && !hasRegion) {
        clientBuilder.region(Region.of(preferredRegion));
      }
      dynamoDbClient = clientBuilder.build();
      return new DynamoDbMetastoreImpl(this);
    }
  }

  public interface EndPointStep {
    /**
     * Adds Endpoint config to the AWS DynamoDb client.
     *
     * @param endPoint The service endpoint either with or without the protocol.
     * @param signingRegion The region to use for SigV4 signing of requests (e.g. us-west-1).
     * @return The current {@code BuildStep} instance with end point configuration.
     */
    BuildStep withEndPointConfiguration(String endPoint, String signingRegion);
  }

  public interface RegionStep {
    /**
     * Specifies the region for the AWS DynamoDb client.
     *
     * @param region The region for the DynamoDb client.
     * @return The current {@code BuildStep} instance with a client region. This overrides the region specified as the
     *         {@code newBuilder} input parameter.
     */
    BuildStep withRegion(String region);
  }

  public interface BuildStep {
    /**
     * Specifies the name of the table.
     *
     * @param tableName A custom name for the table.
     * @return The current {@code BuildStep} instance.
     */
    BuildStep withTableName(String tableName);

    /**
     * Specifies whether key suffix should be enabled for DynamoDB. This should be enabled for Global Table support.
     * Adding a suffix to keys prevents multi-region writes from clobbering each other and ensures that no keys are
     * lost.
     *
     * @return The current {@code BuildStep} instance.
     */
    BuildStep withKeySuffix();


    /**
     * Specifies a custom DynamoDb client to be used instead of the default. Custom client will ignore any region
     * or endpoint configuration specified in the builder.
     *
     * @param client A custom DynamoDb client.
     * @return The current {@code BuildStep} instance.
     */
    BuildStep withClientOverride(DynamoDbClient client);

    /**
     * Builds the finalized {@code DynamoDbMetastoreImpl} with the parameters specified in the builder.
     *
     * @return The fully instantiated {@code DynamoDbMetastoreImpl}.
     */
    DynamoDbMetastoreImpl build();
  }
}
